toolchain/lex/lex.cpp

```diff
@@ -1342,11 +1349,29 @@ auto Lexer::LexFileEnd(llvm::StringRef source_text, ssize_t position) -> void {
   NoteWhitespace();
 
   LexToken(TokenKind::FileEnd, position);
+}
 
+auto Lexer::Finalize() -> void {
   // If we had any mismatched brackets, issue diagnostics and fix them.
   if (has_mismatched_brackets_ || !open_groups_.empty()) {
     DiagnoseAndFixMismatchedBrackets();
   }
+
+  // Reject source files with so many tokens that we may have exceeded the
+  // number of bits in `token_payload_`.
+  //
+  // Note that we rely on this check also catching the case where there are too
+  // many identifiers to fit an `IdentifierId` into a `token_payload_`, and
+  // likewise for `IntId` and so on. If we start adding any of those IDs prior
+  // to lexing, we may need to also limit the number of identifiers etc. here.
```

Comment:
```suggestion
  // to lexing, we may need to also limit the number of IDs here.
```

The period "etc." throws my reading, maybe we can back-reference "any of those IDs"?

Reply:
Maybe even more explicitly:
```suggestion
  // to lexing, we may need to also limit the number of those IDs here.
```